File: src/main/java/com/codingforcookies/betterrecords/common/crafting/recipe/RecipeRecordRepeatable.java
package com.codingforcookies.betterrecords.common.crafting.recipe;

import com.codingforcookies.betterrecords.item.ItemRecord;
import net.minecraft.init.Items;
import net.minecraft.inventory.InventoryCrafting;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.IRecipe;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;

public class RecipeRecordRepeatable implements IRecipe {

    @Override
    public boolean matches(InventoryCrafting inventoryCrafting, World world) {
        ItemStack record = null;
        boolean comparator = false;

        for(int i = 0; i < inventoryCrafting.getSizeInventory(); ++i) {
            ItemStack itemstack = inventoryCrafting.getStackInSlot(i);
            if(itemstack != null) {
                if(itemstack.getItem() instanceof ItemRecord && itemstack.getTagCompound() != null)
                    if(record != null)
                        return false;
                    else
                        record = itemstack;
                else if(itemstack.getItem() == Items.COMPARATOR)
                    if(comparator)
                        return false;
                    else
                        comparator = true;
                else
                    return false;
            }
        }

        return record != null && comparator;
    }

    @Override
    public ItemStack getCraftingResult(InventoryCrafting inventoryCrafting) {
        ItemStack record = null;
        boolean comparator = false;

        for(int i = 0; i < inventoryCrafting.getSizeInventory(); i++) {
            ItemStack itemstack = inventoryCrafting.getStackInSlot(i);
            if(itemstack != null) {
                if(itemstack.getItem() instanceof ItemRecord && itemstack.getTagCompound() != null)
                    if(record != null)
                        return null;
                    else
                        record = itemstack;
                else if(itemstack.getItem() == Items.COMPARATOR)
                    if(comparator)
                        return null;
                    else
                        comparator = true;
                else
                    return null;
            }
        }

        if(record == null || !comparator)
            return null;
        else{
            ItemStack newRecord = ItemStack.copyItemStack(record);

            if(newRecord.getTagCompound() == null)
                newRecord.setTagCompound(new NBTTagCompound());

            newRecord.getTagCompound().setBoolean("repeat", true);

            return newRecord;
        }
    }

    @Override
    public int getRecipeSize() {
        return 10;
    }

    @Override
    public ItemStack getRecipeOutput() {
        return null;
    }

    @Override
    public ItemStack[] getRemainingItems(InventoryCrafting inv) {
        inv.clear();
        return new ItemStack[] {};
    }
}

File: src/main/kotlin/com/codingforcookies/betterrecords/crafting/recipe/RecipeRecordRepeatable.kt
package com.codingforcookies.betterrecords.crafting.recipe

import com.codingforcookies.betterrecords.item.ItemRecord
import com.codingforcookies.betterrecords.item.ModItems
import net.minecraft.init.Items
import net.minecraft.inventory.InventoryCrafting
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.IRecipe
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.NonNullList
import net.minecraft.world.World
import net.minecraftforge.registries.IForgeRegistryEntry

class RecipeRecordRepeatable : IForgeRegistryEntry.Impl<IRecipe>(), IRecipe {

    init {
        setRegistryName("reciperecordrepeatable")
    }

    override fun canFit(width: Int, height: Int) = width * height >= 2

    override fun matches(inventoryCrafting: InventoryCrafting, worldIn: World?): Boolean {
        var foundRecord = false
        var foundComparator = false

        (0 until inventoryCrafting.sizeInventory)
                .map { inventoryCrafting.getStackInSlot(it) }
                .filter { !it.isEmpty }
                .forEach {
                    if (it.item is ItemRecord && !foundRecord) {
                        foundRecord = true
                    } else if (it.item == Items.COMPARATOR && !foundComparator) {
                        foundComparator = true
                    } else {
                        return false
                    }
                }

        return foundRecord && foundComparator
    }

    override fun getCraftingResult(inventoryCrafting: InventoryCrafting): ItemStack? {
        val record =
                (0 until inventoryCrafting.sizeInventory)
                        .map { inventoryCrafting.getStackInSlot(it) }
                        .filter { !it.isEmpty }
                        .find { it.item is ItemRecord }

        return record?.copy()?.apply {
            if (!hasTagCompound()) {
                tagCompound = NBTTagCompound()
            }
            tagCompound!!.setBoolean("repeat", true)
        }
    }

    override fun getRecipeOutput() = ItemStack(ModItems.itemRecord)

    override fun getRemainingItems(inv: InventoryCrafting): NonNullList<ItemStack> {
        inv.clear()
        return NonNullList.create()
    }
}
